### DISCUSSION TITLE
RTK cli generator

### DISCUSSION THREAD

**Discussion by DanielFGray** [reduxjs/redux-toolkit#1285] [Ideas]
I was thinking it'd be really awesome to have an official cli that can do things like

- generating a basic store configuration (with or without TS)
- scaffold a named slice and "install it" to the reducers
- generate an RTK-Query "api slice" and install middleware

I think there's a lot of potential to make some really cool tools for Redux, and potentially raise the bar for React libraries/tools, so I thought I'd start the conversation here.

I'm currently looking into [jscodeshift](https://github.com/facebook/jscodeshift) to see how helpful it would be, I'm sure there's other tools worth looking into that would be helpful?

**Comment by markerikson** [comment]
I know that `nx` does stuff like that: https://nx.dev/latest/angular/getting-started/nx-cli

tbh CLIs generating code have never had any real appeal for me, but clearly I'm not the only person with an opinion here :)

**Reply by phryneas** [reply]
I'm pretty much with Mark with this one. 

Two of these three (store configuration and api creation) are things that should be done only once per application. And they have a lot of nuance in how they can be calibrated, if you want TypeScript or not, if you want to add persistance or not. 

Adding a slice might happen a bit more often, but it is not a lot work either - and stuff like exporting actions can only be done after you added the reducers later.

Honestly, I'd rather prefer optimizing the docs for copy-pastability if we are still missing that in any way than adding a cli tool that has to be super-flexible to project structures and adds all kind of maintenance on top of what we already have.

**Reply by DanielFGray** [reply]
> optimizing the docs for copy-pastability

this is also a great idea, I just can't help thinking it would be a much nicer experience to have a cli tool that does it for the user and wouldn't require switching to a browser.

Store configuration only happens once, but I still see many folks that miss the section in the docs about adding things like
```js
export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
```

Whether they are one-off tasks or repeated, having a cli tool that generates Redux' documented best practices would much more easily guide users (especially the newer ones) into the pit of success.

I have seen a few Redux cli tools but none of them are leveraging RTK, I believe having an "official" cli would only help increase the adoption of RTK and possibly Redux more generally, and hopefully quell/ squash the few remaining "Redux boilerplate" arguments. I'm happy to lead the work (though it might take me a while on my own), but I don't think it would have nearly the same beneficial effects if it is published as a third-party library.

**Reply by markerikson** [reply]
Would it be easier to provide a plugin for some other CLI tool rather than trying to create one from scratch?

**Reply by DanielFGray** [reply]
> Would it be easier to provide a plugin for some other CLI tool rather than trying to create one from scratch?

I would imagine so, and that's fine with me assuming it checks off all the usability boxes (whatever they might be defined as). I don't have any real opinions about the implementation details or specifics of the APIs.

**Reply by markerikson** [reply]
I think this falls under the heading of "things we're not going to try ourselves, but if you want to try it go right ahead, and if you get something working we can figure things out from there" :)

**Reply by philipjc** [reply]
Has anyone started anything?
